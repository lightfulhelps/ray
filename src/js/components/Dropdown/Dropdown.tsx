// import React, { useState, useEffect } from 'react';
// import classNames from 'classnames';

// type Props = {
//   className?: string;
//   isBlock?: boolean;
//   render: (arg0: boolean, arg1: () => void) => React.ReactNode;
// };

// const Dropdown: React.FC<Props> = ({ className, isBlock, render, ...other }) => {
//   const node = null;

//   const [isOpen, setIsOpen] = useState(false);

//   const handleDocumentClick = (e: Event) => {
//     if (!e || (e && !e.target) || (node && node.contains(e.target as any) && node !== e.target)) {
//       return;
//     }

//     setIsOpen(false);
//   };
//   useEffect(() => {
//     document.addEventListener('click', handleDocumentClick, true);

//     return (): void => {
//       document.removeEventListener('click', handleDocumentClick, true);
//     };
//   }, []);

//   const handleToggle = () => {
//     setIsOpen(!isOpen);
//   };

//   const classes = classNames(className, 'dropdown', isBlock ? 'd-block' : 'd-inline-block');

//   return (
//     <div
//       {...other}
//       className={classes}
//       ref={node => {
//         node = node;
//       }}
//     >
//       {render(isOpen, handleToggle)}
//     </div>
//   );
// };

// export default Dropdown;
