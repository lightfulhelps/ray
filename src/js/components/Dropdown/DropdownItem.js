// @flow
import * as React from 'react';
import classNames from 'classnames';
import Icon from '../Icon/Icon';
import type { IconNameType } from '../Icon/icons';

type Props = {
  className?: string,
  icon?: IconNameType,
  label: string,
};

const DropdownItem = ({ className, label, icon, onClick }: Props) => {
  const classes = classNames(className, 'dropdown-item');
  return (
    <a className={classes} href="#" onClick={onClick}>
      <span className="w-100 d-flex align-items-center dropdown-item-content">
        {icon && <Icon name={icon} className="mr-1" />}
        <span>{label}</span>
      </span>
    </a>
  );
};

export default DropdownItem;
