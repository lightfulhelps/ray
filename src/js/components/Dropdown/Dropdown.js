// @flow
import React, { Component } from 'react';
import classNames from 'classnames';
import DropdownItem, { type DropdownItemType } from './DropdownItem';
import Button from '../Button/Button';

type Props = {
  buttonSize?: 'lg' | 'md' | 'sm',
  buttonTheme?: string,
  className?: string,
  footer?: string,
  items: DropdownItemType[],
  position?: 'left' | 'right',
  theme?: string,
};

type State = {
  isOpen: boolean,
};

class Dropdown extends Component<Props, State> {
  state = {
    isOpen: false,
  };

  handleDropdown() {
    this.setState({ isOpen: !this.state.isOpen });
  }

  render() {
    const {
      buttonSize = 'sm',
      buttonTheme = 'light',
      className,
      footer,
      items,
      position = 'left',
      theme = 'dark',
    } = this.props;
    const classes = classNames(className, 'dropdown', `dropdown-${theme}`, 'd-inline-block');

    return (
      <div className={classes}>
        <Button
          onClick={() => this.handleDropdown()}
          icon="menu"
          aria-haspopup="true"
          aria-expanded={this.state.isOpen}
          size={buttonSize}
          theme={buttonTheme}
        />
        <div
          className={classNames('dropdown-menu', `dropdown-menu-${position}`, {
            show: this.state.isOpen,
          })}
          aria-labelledby="dropdownMenuButton"
        >
          {items.map((item, i) => (
            <DropdownItem key={i} label={item.label} icon={item.icon} onClick={item.onClick} />
          ))}
          {footer && <div className="dropdown-footer">{footer}</div>}
        </div>
      </div>
    );
  }
}

export default Dropdown;
