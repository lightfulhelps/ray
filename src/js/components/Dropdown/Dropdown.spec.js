import React from 'react';
import merge from 'lodash/merge';
import { shallow } from 'enzyme';
import Dropdown from './Dropdown';

const setup = (overrides = {}) => {
  const props = merge(
    {
      items: [
        {
          label: 'edit',
          icon: 'edit',
        },
        {
          label: 'delete',
          icon: 'delete',
        },
      ],
    },
    overrides
  );
  const wrapper = shallow(<Dropdown {...props} />);

  return {
    wrapper,
    props,
    dropdownItem: wrapper.find('DropdownItem'),
  };
};

describe('<Dropdown />', () => {
  it('should render', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should NOT render a footer if footer prop is NOT provided', () => {
    const { wrapper } = setup();

    expect(wrapper.find('.dropdown-footer').exists()).toBe(false);
  });

  it('should render a footer if footer prop is provided', () => {
    const { wrapper } = setup({ footer: 'Test' });

    expect(wrapper.find('.dropdown-footer').exists()).toBe(true);
    expect(wrapper.find('.dropdown-footer').text()).toEqual('Test');
  });

  it('should handle the position prop', () => {
    const { wrapper } = setup();

    expect(wrapper.find('.dropdown-menu').hasClass('dropdown-menu-left')).toBe(true);

    wrapper.setProps({ position: 'right' });

    expect(wrapper.find('.dropdown-menu').hasClass('dropdown-menu-right')).toBe(true);
  });
});
