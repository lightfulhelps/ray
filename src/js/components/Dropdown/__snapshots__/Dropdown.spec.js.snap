// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should render 1`] = `
<div
  className="dropdown dropdown-dark d-inline-block"
>
  <Button
    aria-expanded={false}
    aria-haspopup="true"
    icon="menu"
    onClick={[Function]}
    size="sm"
    theme="light"
  />
  <div
    aria-labelledby="dropdownMenuButton"
    className="dropdown-menu dropdown-menu-left"
  >
    <DropdownItem
      icon="edit"
      key="0"
      label="edit"
    />
    <DropdownItem
      icon="delete"
      key="1"
      label="delete"
    />
  </div>
</div>
`;
